## Project 3
#### Name Gender Predictor

##### Maxfield Raynolds
##### Data 620
##### Spring 2026

Import necessary packages and name list.

In [2]:
import nltk
import random
from nltk.corpus import names

Create a list of labeled names.

In [3]:
labeled_names = ([(name, 'male') for name in names.words('male.txt')] + [(name, 'female') for name in names.words('female.txt')])

Shuffle the name list.

In [23]:
random.seed(1234)
random.shuffle(labeled_names)

#### Possible variables in a name:

To create a name gender predictor, it will be necessary to enter features that will aid in a model being more successful at predicting the gender commonly associated with a name. In order to execute this task, it is worthwhile to narrow down possible variables that an algorithm could use. Below is a list of possible variables:

- Name length
- Starting letters (1 to n letters from the beginning of the name)
- Ending letters (1 to n letters from the end of the name)
- Specific letters
- Count of letters
- Specific position of letters
- Specific grouping of consonants (as in do specific sets of consonants play a role)

While this list is not exhaustive, it covers many possibilities.

#### Testing the first variable

Below, over the course of three code blocks, the first variable: length, will be tested.

Create a function to start testing variables:

In [24]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)

    return features

Then we will create the featuresets:

In [25]:
featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

Finally, we will instantiate the NaiveBayesClassifer, train it and test it against the dev_test_set.

In [26]:
classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.654

This results in a 65.4% accuracy, or just under 2/3rds correct. There is clearly room for improvement. Below, the featureset will be expanded and various iterations run to test the accuracy. The three previous code blocks will be combined for efficiency.

First is to check the addition of the last letter.

In [27]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    #features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.784

This improves the prediction significantly now to 78.4%. But is length adding to this? Or is it overfitting. Below is just the last letter evaluated:

In [28]:
def gender_features(name):
    features = {}
    #features['name_length'] = len(name)
    features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    #features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.784

At 78.4%, it turns out length is making now difference. Now what if we include the first letter:

In [30]:
def gender_features(name):
    features = {}
    # features['name_length'] = len(name)
    features['last_letter'] = name[-1]
    features['first_letter'] = name[0]
    #features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.804

Interestingly at 80.4%, this is better than when we had not included the first letter. Let's add in the last two letters:

In [31]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.78

At 78%, this is the slightly worse. Could considering both the last letter and the last two letters as independent features be limiting the model? Let's see:

In [34]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.806

Using only the length and the last two letters yields an 80.6% accuracy. While the first letter was not helpful, perhaps the first two would be useful.

In [35]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    features['first_two_letters'] = name[:2]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.806

As it turns out, at 80.6% the first two letters have not made an improvement.  Let's now include, a count of each letter used in each name:

In [38]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    for letter in 'abcdefghijklmnopqrstuvwxyz':
        features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.826

This was a bit useful. How about whether a letter exists in the name:

In [40]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #   features['has({})'.format(letter)] = (letter in name.lower())
    #    features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.826

This made no difference. Finally, let's see if a letter position helps:

In [41]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    features['first_two_letters'] = name[:2]
    for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
        features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.814

Again this is not particularly helpful. So the best result on the dev_test_set has been a combination of:

- length
- the first two letters
- the last two letters

What if we consider more letters sucha as the first three and last three:

In [43]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    features['first_two_letters'] = name[:2]
    features['last_three_letters'] = name[-3:]
    features['first_three_letters'] = name[:3]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #   features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.844

The best result yet. Interestingly, this result does not decline if the first two letters are removed as a feature:


In [44]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    features['last_three_letters'] = name[-3:]
    features['first_three_letters'] = name[:3]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #   features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, dev_test_set)

0.854

The final results of the testing indicate that the following features are useful in the name gender predictor:
- name length
- the last two letters
- the last three letters
- the first three letters

But how will the model perform on the reserved test set?


In [45]:
def gender_features(name):
    features = {}
    features['name_length'] = len(name)
    #features['last_letter'] = name[-1]
    #features['first_letter'] = name[0]
    features['last_two_letters'] = name[-2:]
    #features['first_two_letters'] = name[:2]
    features['last_three_letters'] = name[-3:]
    features['first_three_letters'] = name[:3]
    #for letter in 'abcdefghijklmnopqrstuvwxyz':
    #    features['count({})'.format(letter)] = name.lower().count(letter)
    #    features['has({})'.format(letter)] = (letter in name.lower())
    #   features['place({})'.format(letter)] = name.lower().find(letter)

    return features

featuresets = [(gender_features(n), gender) for (n,gender) in labeled_names]
test_set, dev_test_set, train_set = featuresets[:500], featuresets[500:1000], featuresets[1000:]

classifier = nltk.NaiveBayesClassifier.train(train_set)
nltk.classify.accuracy(classifier, test_set)

0.852

The result on the reserved test set is 85.2%. This is lower than the dev_test_set, but not by much. The reduction is likely going to be a result of slight overfitting to the specific training set, its specific randomization, and the specific names separated as the test set. That being said 85.2% is still getting more than 4 out of 5 genders correct.

Below is an examination of the most useful features:

In [46]:
classifier.show_most_informative_features(20)

Most Informative Features
        last_two_letters = 'na'           female : male   =     93.3 : 1.0
        last_two_letters = 'us'             male : female =     37.2 : 1.0
        last_two_letters = 'ia'           female : male   =     37.2 : 1.0
        last_two_letters = 'sa'           female : male   =     32.6 : 1.0
      last_three_letters = 'nne'          female : male   =     31.5 : 1.0
        last_two_letters = 'io'             male : female =     25.8 : 1.0
        last_two_letters = 'ra'           female : male   =     25.4 : 1.0
        last_two_letters = 'rd'             male : female =     24.7 : 1.0
        last_two_letters = 'ta'           female : male   =     23.7 : 1.0
      last_three_letters = 'tta'          female : male   =     23.3 : 1.0
      last_three_letters = 'ana'          female : male   =     22.8 : 1.0
        last_two_letters = 'rt'             male : female =     22.4 : 1.0
        last_two_letters = 'ld'             male : female =     21.4 : 1.0

According to the first twenty most informative features, it was all the last two or three letters that are most informative. Additionally, the split is almost even as to the most informative features between male and female.